Format resolving schema conflicts lesson (#107)
Ada-Developers-Academy/core-unit-2: api-3-database-models-read/resolving-schema-conflicts.md

### !callout-info

### Source: Miguel Grinberg's Blog

This solution and explanation to resolving schema conflicts is taken from [Miguel Grinberg's blog](https://blog.miguelgrinberg.com/post/resolving-database-schema-conflicts). 
</br>
</br>
We made a few minor changes to the wording, we changed the example model from `User` to `Book`, and we changed the names of the two developers to Ada instructors.

### !end-callout

## Goals

Our goal for this lesson to describe one method for resolving database schema conflicts.

## Introduction

If you work on a project that uses database migrations with other developers, it is likely that you have experienced migration conflicts at some point. These occur when two or more developers are merging unrelated features to the master source control branch at around the same time, with each feature requiring different changes to the database.

### !callout-info

## Avoid Conflicts

No matter how careful we are, we will encounter conflicts. 

However, here are a few tips to help avoid conflicts in the database schema:
* `git pull` often
  * especially, `git pull` before applying changes to the model with `flask db migrate`

> * Work on updates to the models together
>   * Even if working on separate models, conflicts to the database schema can arise

Reviewer: ```suggestion
* Work on updates to the models together
  * We strongly recommend using pair programming for every change you make to a model in a group project.
  * Even if working on separate models, conflicts to the database schema can arise.
```

    author = db.Column(db.String)
```

Immediately after editing the model, she generates and applies a database migration, to make this change on her development database:

```
$ flask db migrate -m "add author to book"
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.ddl.postgresql] Detected sequence named 'book_id_seq' as owned by integer column 'book(id)', assuming SERIAL and omitting
INFO  [alembic.autogenerate.compare] Detected added column 'book.author'
  Generating /hello-books-
  api/migrations/versions/bf69b044cdfc_add_author_to_book.py ...  done

$ flask db upgrade
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade f9e86c06ab0d -> bf69b044cdfc, add author to book
```

She then checks the database migration history to make sure the change was applied:

```
$ flask db history
f9e86c06ab0d -> bf69b044cdfc (head), add author to book
<base> -> f9e86c06ab0d, add book table
```

So Audrey happily goes off to work on adding authors.

> Meanwhile, Trenisha needs to add an isbn number for all the books, so starting from the same Book model as Audrey, she makes the following change on his development environment:

Reviewer: ```suggestion
Meanwhile, Trenisha needs to add an ISBN (a number used to track book barcodes) for all the books, so starting from the same Book model as Audrey, she makes the following change on her development environment:
```

```python
class Book(db.Model):
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    title = db.Column(db.String)
    description = db.Column(db.String)
    isbn = db.Column(db.Integer)
```

And she also creates a database migration for this change:

```
$ flask db migrate -m "add isbn to book"
INFO  [alembic.autogenerate.compare] Detected added column 'book.isbn'
  Generating migrations/versions/73c1f8470b04_add_isbn_to_book.py ... done

$ flask db history
f9e86c06ab0d -> 73c1f8470b04 (head), add isbn to book
<base> -> f9e86c06ab0d, add book table
```

> Let's assume that Trenisha finishes his work first, and pushes the user avatar changes to the upstream repository. This commit triggers an automatic deployment to a staging server, where Trenisha goes and checks that user avatars are working fine. She finds no problems, so she goes to find other work.

Reviewer: ```suggestion
Let's assume that Trenisha finishes her work first, and pushes the book changes to the upstream repository. This commit triggers an automatic deployment to a staging server, where Trenisha goes and checks that ISBNs are working fine. She finds no problems, so she goes to find other work.
```

> Let's assume that Trenisha finishes his work first, and pushes the user avatar changes to the upstream repository. This commit triggers an automatic deployment to a staging server, where Trenisha goes and checks that user avatars are working fine. She finds no problems, so she goes to find other work.

Reviewer: Here and elsewhere, will talking about a "staging server" be confusing? I don't think we've talked about this concept before.

Author: I agree this could be confusing. What do you think about simply removing that line

```suggestion
Let's assume that Trenisha finishes his work first, and pushes the user avatar changes to the upstream repository.
```

Author: Updated suggestion below with `book author` instead of `user avatar`

f9e86c06ab0d -> 73c1f8470b04 (head), add isbn to book
<base> -> f9e86c06ab0d (branchpoint), add book table
```

So what happened here? As you see above, the migration history is not linear anymore. Migration `f9e86c06ab0d` was the starting migration for both Audrey and Trenisha, so both their migrations are based on this change, creating a branched migration history with two heads.

It's interesting to note that when Audrey pulled Trenisha's change git did not complain. These schema migrations are stored each in its own file, so git did not see conflicts in Audrey's change and allowed her change to be merged, causing the project to break. Not nice, right?

Notice that in Audrey's database, there is no `isbn` column.

```
$ SELECT * FROM book;
 id | title | description | author 
----+-------+-------------+--------
(0 rows)
```

Likewise in Trenisha's database, there is no `author` column.

## How to Detect Schema Conflicts Before They are Committed
Before I tell you how to untangle this schema mess, let's think about Audrey's actions. Could she or her team have done anything different to prevent conflicts like this from ever appearing in the team's repository?

> There are certainly ways to prevent this type of conflicts. A migration history test could be written to find this and other problems with migrations. This is a test that creates an empty database, and simply applies all the migrations in order, to ensure that they all run fine. The test can then downgrade the database all the way back to its initial state, to also test downgrades, something that very few people check. A migration history test can be included as part of the application's unit test suite, or as a source control pre-commit check.

Reviewer: This is too complicated a solution for our students. Let's just reiterate about pulling frequently and using pair programming for any model work.

Author: What do you think about just removing this section? We already have a callout at the beginning about pair programming and pulling frequently.

Author: suggestion removing section below

> There are certainly ways to prevent this type of conflicts. A migration history test could be written to find this and other problems with migrations. This is a test that creates an empty database, and simply applies all the migrations in order, to ensure that they all run fine. The test can then downgrade the database all the way back to its initial state, to also test downgrades, something that very few people check. A migration history test can be included as part of the application's unit test suite, or as a source control pre-commit check.

Reviewer: Agree on deleting this

INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade f9e86c06ab0d -> bf69b044cdfc, add author to book
```

She then checks the database migration history to make sure the change was applied:

```
$ flask db history
f9e86c06ab0d -> bf69b044cdfc (head), add author to book
<base> -> f9e86c06ab0d, add book table
```

So Audrey happily goes off to work on adding authors.

Meanwhile, Trenisha needs to add an ISBN for all the books, so starting from the same Book model as Audrey, she makes the following change on her development environment:

```python
class Book(db.Model):
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    title = db.Column(db.String)
    description = db.Column(db.String)
    isbn = db.Column(db.Integer)
```

And she also creates a database migration for this change:

```
$ flask db migrate -m "add isbn to book"
INFO  [alembic.autogenerate.compare] Detected added column 'book.isbn'
  Generating migrations/versions/73c1f8470b04_add_isbn_to_book.py ... done

$ flask db history
f9e86c06ab0d -> 73c1f8470b04 (head), add isbn to book
<base> -> f9e86c06ab0d, add book table
```

> Let's assume that Trenisha finishes her work first, and pushes the book changes to the upstream repository. This commit triggers an automatic deployment to a staging server, where Trenisha goes and checks that ISBNs are working fine. She finds no problems, so she goes to find other work.

Author: Suggest removing references to staging server

```suggestion
Let's assume that Trenisha finishes her work first, and pushes the book changes to the upstream repository. 
```

> When Audrey completes her work on user authentication, she tries to push to master and gets an error that tells her that her source tree is out of date. So she does a `git pull`, and then tries to push again. This time the push succeeds, so then she anxiously waits for the staging server to update so that she can check her work before moving on. But something bad happened, the deployment to the staging server failed horribly.
> 
> Can you see why? When Audrey goes to check the logs of the failed deployment, this is what she finds:

Author: `flask db upgrade` isn't checking the logs. I think when we replicated this it showed up a little differently. How does this statement look for representing what we saw?

```suggestion
When Audrey completes her work on the book author, she tries to push to master and gets an error that tells her that her source tree is out of date. So she does a `git pull`, and then tries to push again. This time the push succeeds. To apply to new migrations she runs `flask db upgrade`, and sees that it fails silently:
```

Reviewer: Looks good to me

```
$ flask db upgrade
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
```

> It silently fails.
> 
> And sure enough, she did not realize this, but even her own source tree got into a weird state after she pulled in Trenisha's change:

Author: ```suggestion
Sure enough, she did not realize this, but even her own source tree got into a weird state after she pulled in Trenisha's change:
```

Notice that in Audrey's database, there is no `isbn` column.

```
$ SELECT * FROM book;
 id | title | description | author 
----+-------+-------------+--------
(0 rows)
```

Likewise in Trenisha's database, there is no `author` column.

## How to Detect Schema Conflicts Before They are Committed
Before I tell you how to untangle this schema mess, let's think about Audrey's actions. Could she or her team have done anything different to prevent conflicts like this from ever appearing in the team's repository?

There are certainly ways to prevent this type of conflicts. A migration history test could be written to find this and other problems with migrations. This is a test that creates an empty database, and simply applies all the migrations in order, to ensure that they all run fine. The test can then downgrade the database all the way back to its initial state, to also test downgrades, something that very few people check. A migration history test can be included as part of the application's unit test suite, or as a source control pre-commit check.

## How to Resolve a Schema Conflict with a Merge

> While detecting these conflicts before they are pushed to the shared repository is important, let's not forget about Audrey's situation. Because she applied her migration before pulling in Trenisha's changes, her database is aligned with one of the two branched heads in the migration history. Trenisha's migration is in the other branch, so it can't be applied through an upgrade.

Author: ```suggestion
Let's help Audrey with her situation. Because she applied her migration before pulling in Trenisha's changes, her database is aligned with one of the two branched heads in the migration history. Trenisha's migration is in the other branch, so it can't be applied through an upgrade.
```
